src/io.rs

```diff
@@ -0,0 +1,102 @@
+//! Simple abstractions for the IO operations.
+//!
+//! Basically it just a replacement for the std::io that is usable from
+//! the `no_std` environment.
+
+use std::vec::Vec;
+
+/// IO specific error.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// Some unexpected data left in the buffer after reading all data.
+	TrailingData,
+
+	/// Unexpected End-Of-File
+	UnexpectedEof,
+
+	/// Invalid data is encountered.
+	InvalidData,
+}
+
+/// IO specific Result.
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+pub trait Write {
+	/// Write a buffer of data into this write.
+	///
+	/// All data is written at once.
+	fn write(&mut self, buf: &[u8]) -> Result<()>;
+}
+
+pub trait Read {
+	/// Read a data from this read to a buffer.
+	///
+	/// If there is not enough data in this read then `UnexpectedEof` will be returned.
+	fn read(&mut self, buf: &mut [u8]) -> Result<()>;
```

Comment:
Can we change this to read_exact to keep the semantics the same with `std::io`? Or maybe just implement a `no_std` version of `Read::read_exact` and keep the unimplemented function as it is originally `Read::read`?

Reply:
As discussed with @sorpaas: this isn't such a big concern given that `Read`/`Write` are private.